### PR TITLE
frozendict should be covariant on V, like Mapping.

### PR DESCRIPTION
**Pull request by asford**
Immutable collections, like `Sequence`, `tuple` and `frozendict` are [covariant] in their contained types.
The `frozendict` base interface, `Mapping[K, V]`, is covariant in  `V`.

Update the `frozendict` .pyi stubs to also specify a covariant `V` for `frozendict[K,V]`.
This enables us to assign collections-of-subtypes to frozendict generics,
which currently raises a type error to value type invariance.

For example:

```python
string_vals: frozendict[str, str] = frozendict(a="foo")
vals: frozendict[str, Hashable] = string_vals # should be allowed
```

Currently raises a type error:

```
Expression of type "frozendict[str, str]" cannot be assigned to declared type "frozendict[str, Hashable]"
  "frozendict[str, str]" is incompatible with "frozendict[str, Hashable]"
    Type parameter "V@frozendict" is invariant, but "str" is not the same as "Hashable"
```

but is a valid assignment.

[covariant]: https://www.playfulpython.com/type-hinting-covariance-contra-variance/#:~:text=A%20data%20type%20is%20covariant,the%20parent%20class%20is%20expected.

### CODE REVIEW

**Comment by asford** [comment]
@Marco-Sulla this is a minor fix for the stubs added in #70.

Thanks for the great work on this project.

**Comment by Marco-Sulla** [comment]
`dict` is invariant too.

```
from collections.abc import Hashable
string_vals: dict[str, str] = dict(a="foo")
vals: dict[str, Hashable] = string_vals
```

mypy output:

test/typed.py:38: error: Incompatible types in assignment (expression has type "Dict[str, str]", variable has type "Dict[str, Hashable]")  [assignment]
test/typed.py:38: note: "Dict" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
test/typed.py:38: note: Consider using "Mapping" instead, which is covariant in the value type

**Comment by asford** [comment]
dict implements MutableMapping, as it's a mutable container and is therefore invariant.

frozendict is a somewhat special case, as it implements the (covariant) Mapping interface.

By analogy, frozendict is to dict as tuple is to list or frozenset is to set. It's a (essentially covariant) immutable container.

I think you've extended this with non-covariant-ish update/set/__ror__ methods, however this can be appropriately typed using a non-covariant V2 parameter returning a union type, as you've done here, because it does not modify the source frozendict.

Would you be open to reopening with updates to allow covariant behavior on V, and non-covariant extensions? This is actually a pretty useful behavior for frozen dataclasses where we may want to have a concrete, dict-like property.

**Comment by Marco-Sulla** [comment]
> dict implements MutableMapping, as it's a mutable container and is therefore invariant.

Mh, you're right. This works:

```
from collections.abc import Hashable
string_vals: tuple[str] = ("foo", )
vals: tuple[Hashable] = string_vals
```

For now I simply reopen this, but I check it more deeply later.

**Comment by Marco-Sulla** [comment]
(Personal rant: Python typing is a mess IMHO, and I repeat IMHO. The syntax is ugly, is quite difficult to implement and to debug.)

**Comment by Marco-Sulla** [comment]
Could you also add this to `test/typed.py`?

```
from collections.abc import Hashable
string_vals: frozendict[str, str] = frozendict(a="foo")
vals: frozendict[str, Hashable] = string_vals
```

**Comment by asford** [comment]
> (Personal rant: Python typing is a mess IMHO, and I repeat IMHO. The syntax is ugly, is quite difficult to implement and to debug.)

I definitely ...don't disagree.... 😉 

Thanks for pointing me to `test/typed.py`. I took the liberty of:
* Updating `test/typed.py` file to use `assert_type`, which documents the expected type of a statement and verifies this type via the checker.
* Adding typing checks with `pyright` and `mypy` (if installed) to mytest.sh

In doing so, I found a few small bugs in `__init__.pyi`. I've updated these type signatures to clarify the covariance issues we described above as well as tightened up the type declarations for the extended `set` and `setdefault` methods.

This fixes a few errors I was seeing in type inference under `pylance` in vscode.

File: src/frozendict/__init__.pyi [hunk 1]
```diff
@@ -1,73 +1,52 @@
 from __future__ import annotations
 
-from typing import TypeVar, overload, Optional
+from typing import Optional, TypeVar, Union, overload
```
Comment (Marco-Sulla) [inline src/frozendict/__init__.pyi:3]: Ah, Pycharm here :D

**Comment by asford** [comment]
Fantastic, thanks for the review. 

I'm happy to setup a quick tox-based test config that includes your existing suite and these type tests if you'd like. 

It might help if you have other contributors come by and could be run in actions.

**Comment by Marco-Sulla** [comment]
> It might help if you have other contributors

I can contribute, but please, let me know about the open points, especially the ones about `__init__.pyi`

**Comment by asford** [comment]
I'm not sure I understand? There aren't any review comments open right now. The test suite with type assertions is passing on this branch w/ mypy and pyright.

**Comment by Marco-Sulla** [comment]
> There aren't any review comments open right now.

Emh. Can you scroll this discussion up? :D

**Comment by Marco-Sulla** [comment]
In particular all the "Why you removed" etc

**Comment by asford** [comment]
> In particular all the "Why you removed" etc

I don't see any review comments, maybe they're in draft?

**Review by Marco-Sulla — CHANGES_REQUESTED** [review]

**Comment by Marco-Sulla** [comment]
Mh maybe I forgot to submit review...

**Comment by Marco-Sulla** [comment]
Superseeded by https://github.com/Marco-Sulla/python-frozendict/commit/ebbc70b20b434b364fb9897dfe54cf7badacf700 and subsequent commits.

**Pull request closed without merging**